sdk/go/pulumi/policyx/policyPack.go

```diff
@@ -0,0 +1,125 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policyx
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumix"
+)
+
+type HandshakeRequest struct {
+	// The "host" of the analyzer. This is the engine that is running the analyzer.
+	Host pulumix.Engine
+	// A *root directory* where the analyzer's binary, `PulumiPolicy.yaml`, or other identifying source code is located.
+	// In the event that the analyzer is *not* being booted by the engine (e.g. in the case that the engine has been
+	// asked to attach to an existing running analyzer instance via a host/port number), this field will be empty.
+	RootDirectory *string
+	// A *program directory* in which the analyzer should execute. This is generally a subdirectory of the root
+	// directory, though this is not required. In the event that the analyzer is *not* being booted by the engine (e.g.
+	// in the case that the engine has been asked to attach to an existing running analyzer instance via a host/port
+	// number), this field will be empty.
+	ProgramDirectory *string
+}
+
+type HandshakeResponse struct{}
+
+// PolicyConfig provides configuration for a policy.
+type PolicyConfig struct {
+	// Enforcement level of the policy.
+	EnforcementLevel EnforcementLevel
+	// Configuration properties of the policy.
+	Properties map[string]any
+}
+
+// ConfigureRequest provides configuration information to the analyzer.
+type ConfigureRequest struct {
+	// Map of policy name to config.
+	PolicyConfig map[string]PolicyConfig
+}
+
+type PolicyPack interface {
+	Name() string
+	Version() semver.Version
+	Handshake(context.Context, HandshakeRequest) (HandshakeResponse, error)
+	Policies() []Policy
+}
+
+type policyPack struct {
+	name      string
+	version   semver.Version
+	handshake func(context.Context, HandshakeRequest) (HandshakeResponse, error)
+	policies  []Policy
+}
+
+var policyPackNameRE = regexp.MustCompile(`^[a-zA-Z0-9-_.]{1,100}$`)
+
+func NewPolicyPack(
+	name string, version semver.Version, enforcementLevel EnforcementLevel,
+	handshake func(context.Context, HandshakeRequest) (HandshakeResponse, error),
+	policies []Policy,
+) (PolicyPack, error) {
+	if name == "" || !policyPackNameRE.MatchString(name) {
+		return nil, fmt.Errorf("invalid policy pack name: %q", name)
+	}
+
+	for _, policy := range policies {
+		if policy.Name() == "all" {
+			return nil, fmt.Errorf("invalid policy name %[1]q. %[1]q is a reserved name", policy.Name())
+		}
+
+		schema := policy.ConfigSchema()
+
+		if schema != nil {
+			if _, ok := schema.Properties["enforcementLevel"]; ok {
+				return nil, errors.New("enforcementLevel cannot be explicitly specified in configSchema properties")
```

Comment:
Because it's reserved, I'm guessing? Might be worth augmenting these messages like the one above to say it's reserved -- also clearer for folks coming to this code later on?